ASAN_TRAP in WebCore::LegacyRenderSVGShape::strokeBoundingBox

lopezlozoya opened the pull request:
#### 92b2992e4a8135455e50406ff72fd3b1b86b4f29
<pre>
ASAN_TRAP in WebCore::LegacyRenderSVGShape::strokeBoundingBox
<a href="https://bugs.webkit.org/show_bug.cgi?id=295535">https://bugs.webkit.org/show_bug.cgi?id=295535</a>
<a href="https://rdar.apple.com/153080312">rdar://153080312</a>

Reviewed by Simon Fraser and Ryosuke Niwa.

Adding checks for invalid bounding box leading marker to NaN.

* LayoutTests/fast/svg/path-NaN-stroke-bounding-box-crash-expected.txt: Added.
* LayoutTests/fast/svg/path-NaN-stroke-bounding-box-crash.html: Added.
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
(WebCore::LegacyRenderSVGPath::adjustStrokeBoundingBoxForMarkersAndZeroLengthLinecaps const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
(WebCore::LegacyRenderSVGShape::calculateStrokeBoundingBox const):

Canonical link: <a href="https://commits.webkit.org/297180@main">https://commits.webkit.org/297180@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/28dd1896a90c83f8eb4e8bcb14bc3f0c72623f79

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110786 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30449 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20881 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/116815 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/61054 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112749 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31129 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39035 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/116815 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/61054 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113734 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24862 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/116815 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/24229 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK1-Tests-EWS "Waiting to run tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/60609 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/94255 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17952 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/119605 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37831 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/28115 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/119605 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38207 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96023 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/119605 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23709 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/38060 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/33805 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37724 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/43194 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/37385 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/40723 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/39092 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
```diff
@@ -86,8 +86,12 @@ FloatRect LegacyRenderSVGPath::adjustStrokeBoundingBoxForMarkersAndZeroLengthLin
+        FloatRect markerRect = this->markerRect(repaintRectCalculation, strokeWidth);
+        if (!markerRect.isNaN())
+            strokeBoundingBox.unite(markerRect);
+
```
A superfluous blank line.

smfr commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
```diff
@@ -402,9 +402,11 @@ FloatRect LegacyRenderSVGShape::calculateStrokeBoundingBox() const
             }
         } else {
-            strokeBoundingBox.unite(path().strokeBoundingRect(Function<void(GraphicsContext&)> { [this] (GraphicsContext& context) {
+            FloatRect strokeBoundingRect = path().strokeBoundingRect(Function<void(GraphicsContext&)> { [this] (GraphicsContext& context) {
```
`auto strokeBoundingRect`

smfr commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
```diff
@@ -86,8 +86,12 @@ FloatRect LegacyRenderSVGPath::adjustStrokeBoundingBoxForMarkersAndZeroLengthLin
-    if (!m_markerPositions.isEmpty())
-        strokeBoundingBox.unite(markerRect(repaintRectCalculation, strokeWidth));
+    if (!m_markerPositions.isEmpty()) {
+        FloatRect markerRect = this->markerRect(repaintRectCalculation, strokeWidth);
```
`auto markerRect`

smfr commented on LayoutTests/fast/svg/path-NaN-stroke-bounding-box-crash.html:
```diff
@@ -0,0 +1,19 @@
+<script>
+if (testRunner)
+    testRunner.dumpAsText();
+function jsfuzzer() {
```
Can we make this less like a fuzzer test case? Give this function a normal name, give the variables reasonable names.

rniwa commented on LayoutTests/fast/svg/path-NaN-stroke-bounding-box-crash.html:
```diff
@@ -0,0 +1,15 @@
+<p>This test passes if it doesn't crash.</p>
+<script>
+window?.testRunner?.dumpAsText();
+function createsvg() {
```
I would have called this `createSVGPath` or `createSVGPathSeg`.

lopezlozoya committed:
ASAN_TRAP in WebCore::LegacyRenderSVGShape::strokeBoundingBox
https://bugs.webkit.org/show_bug.cgi?id=295535
rdar://153080312

Reviewed by Simon Fraser and Ryosuke Niwa.

Adding checks for invalid bounding box leading marker to NaN.

* LayoutTests/fast/svg/path-NaN-stroke-bounding-box-crash-expected.txt: Added.
* LayoutTests/fast/svg/path-NaN-stroke-bounding-box-crash.html: Added.
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
(WebCore::LegacyRenderSVGPath::adjustStrokeBoundingBoxForMarkersAndZeroLengthLinecaps const):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
(WebCore::LegacyRenderSVGShape::calculateStrokeBoundingBox const):

Canonical link: https://commits.webkit.org/297180@main
